### ISSUE TITLE
Streaming logs using attach causing hang

### ISSUE THREAD

**Issue by xsresearch** [swipely/docker-api#511]
Streaming logs using the command suggested in the documentation sometimes causes a hang, and the attach function never returns. This appears to occur on very short running containers.

> container.tap(&:start).attach { |stream, chunk| puts "#{stream}: #{chunk}" }

The issue can be replicated by running container.start and then once the container has exited, running the attach command with a block supplied.

A workaround is to run the attach in a separate thread then launch a monitor thread which checks the container status. If container exit is detected, the thread running attach is terminated.

Server version is 17.09.0-ce and kernel version is 3.10.0-514.el7.x86_64. Please ask for further details if necessary to replicate.

**Comment by tlunter** [comment]
@xsresearch this is actually the flow I use in my own project because of short lived containers:

```
    container = Docker::Container.create(...)

    attach_thread = Thread.new do
      container.attach(logs: true, stream: true, stdout: true, stderr: true, tty: true) do |chunk|
        chunk.lines do |line|
          buffers[:stdout].process_line(line)
        end
      end
    end

    attach_thread.join(0.5)

    container.start

    attach_thread.join

    logger.info("Finished executing '#{command}' in local docker")

    exit_code = container.wait['StatusCode'].to_i
```

I don't think there is anything you can do to get around that with docker though. They use goroutines and channels in the docker client which handles this request pattern much better.

**Comment by xsresearch** [comment]
@tlunter thanks for looking at this, relieved I didn't miss any obvious solution.

I did originally write something similar to your suggestion, but got stuck on the delay which you have as 0.5 sec, as if the server is under heavy load the attach process could theoretically take longer then the container run (although practically probably not), and delving deeper I saw no easy way to signal that the attach has completed.

The solution I implemented is in https://github.com/libroute/libroute/commit/9bd861266abe174eb84858b6b1dc8f2ed1869a7b and aims to ensure that a hang is not possible. There still is a delay which waits to ensure that the stream has finished, but worst cast under heavy load should be missing logs rather than a potential hang.

Might be worth updating the readme with a comment and/or suggested solution where it provides this method, as this issue wasn't apparent.
